docs: assign remove function (#192)
tier4/CARET_doc: docs/configuration/intra_node_data_path.md

    `use_latest_message` and `callback_chain` do not cover all of use cases for CARET. We, CARET development team, continue improvements of intra-node data path definition.
<prettier-ignore-end>

## Define intra-node data path to the Architecture file/object

In order to measure latency in the inter-node data path, the Architecture file must be edited to match the entity of the application to be measured.

There are two ways to define inter-node data path.

> 1. Directly editing the Architecture file with an editor, etc

Author: Fixed it in [fb2c5ab](https://github.com/tier4/CARET_doc/pull/192/commits/fb2c5ab09719329bb9546fcdba69c3295f8991f2).

2. Manipulating the Architecture object using the Python API

> Either way will produce the same results as long as it is defined correctly, so define the data path as you prefer.

Reviewer: ```suggestion
Both methods will yield the same results as long as it is defined correctly, so define the data path as you prefer.
```

Author: Fixed it in [fb2c5ab](https://github.com/tier4/CARET_doc/pull/192/commits/fb2c5ab09719329bb9546fcdba69c3295f8991f2).

> ### Directly editing the file

Reviewer: ```suggestion
### Directly editing the Architecture file
```

Author: There is a concern that the names of the sub-sections displayed in the `Table of contents` will become too long.

User have to fill callback name in `variable_passings`, `publishes`'s `callback_name`. `context_type` must be set as `callback_chain`.
After editing, use path.verify() described in the beginning of this section to verify that it has been set correctly.

> ### Manipulating the object

Reviewer: ```suggestion
### Manipulating the Architecture object
```

Author: There is a concern that the names of the sub-sections displayed in the `Table of contents` will become too long.

> CARET serves Python-based APIs to define intra-node data path.

Reviewer: ```suggestion
CARET serves Python API to define intra-node data path.
```

Author: Fixed it in [fb2c5ab](https://github.com/tier4/CARET_doc/pull/192/commits/fb2c5ab09719329bb9546fcdba69c3295f8991f2).

> The following commands enable `Architecture file editing` done in the previous section to be implemented by python commands.

Author: Fixed it in [fb2c5ab](https://github.com/tier4/CARET_doc/pull/192/commits/fb2c5ab09719329bb9546fcdba69c3295f8991f2).

> All of the following code snippets can be executed after the `Architecture('type', 'file_path')` method loads an architecture object.

Reviewer: ```suggestion
All of the code snippets provided below can be executed after loading an architecture object using the `Architecture('type', 'file_path')` method.
```

Author: Fixed it in [9770137](https://github.com/tier4/CARET_doc/pull/192/commits/977013785203771f1c5ce427ab8f2e544a807b88).

The text `All of the following code snippets...` is based on the following pages.

https://tier4.github.io/CARET_doc/latest/configuration/load_and_save/#python-api

Reviewer: OK.

The architecture object whose sub-objects are renamed is saved to a file as explained in [the previous page](./load_and_save.md#save).

#### `use_latest_message`

Using the following Python API to use `use_latest_message` on an architecture object. The examples below follow the previous section.

> - You can add callback to publisher with `insert_publisher_callback` function in `Architecture` class.

Reviewer: ```suggestion
- You can add a callback to a publisher with `insert_publisher_callback` function in `Architecture` class.
```

Author: Fixed it in [fb2c5ab](https://github.com/tier4/CARET_doc/pull/192/commits/fb2c5ab09719329bb9546fcdba69c3295f8991f2).

- You can update `context_types` to `use_latest_message` between targeted subscription and publisher with `update_message_context` function in `Architecture` class.
  - As arguments, the target node name, subscription topic name and publisher topic name must be specified.

```python
# arch is caret_analyze.architecture.architecture.Architecture-based object

arch.insert_publisher_callback('/pong_node', '/pong', 'timer_callback_1')
arch.update_message_context('/pong_node', '/ping', '/pong', 'use_latest_message')
```

As a result of these processes, data path is defined as `use_latest_message`.
The edited architecture file is output by `arch.export()` function.

You can also remove intra-node data path of `use_latest_message` in Architecture object using the following Python API.

```python
# arch is caret_analyze.architecture.architecture.Architecture-based object

arch.remove_publisher_callback('/pong_node', '/pong', 'timer_callback_1')
arch.update_message_context('/pong_node', '/ping', '/pong', 'UNDEFINED')
```

#### `callback_chain`

Using the following Python API to use `callback_chain` on an architecture object. The examples below follow the previous section.

> - You can add callback to publisher with `insert_publisher_callback` function in `Architecture` class.

Author: Fixed it in [fb2c5ab](https://github.com/tier4/CARET_doc/pull/192/commits/fb2c5ab09719329bb9546fcdba69c3295f8991f2).

  - As arguments, the target node name, publishing topic name and publisher callback name must be specified.

- You can add variable passing with `insert_variable_passing` function in `Architecture` class.

  - As arguments, the target node name, write callback name and read callback name must be specified.

- You can update `context_types` to `callback_chain` between targeted subscription and publisher with `update_message_context` function in `Architecture` class.
  - As arguments, the target node name, subscription topic name and publisher topic name must be specified.

```python
arch.insert_publisher_callback('/pong_node', '/pong', 'timer_callback_1')
arch.insert_variable_passing('/pong_node', 'subscription_callback_0', 'timer_callback_1')
arch.update_message_context('/pong_node', '/ping', '/pong', 'callback_chain')
```

As a result of these processes, data path is defined as `callback_chain`.
The edited architecture file is output by `arch.export()` function.

> You can also remove intra-node data path of `callback_chain` in Architecture object using the following Python API.

Author: Fixed it in [fb2c5ab](https://github.com/tier4/CARET_doc/pull/192/commits/fb2c5ab09719329bb9546fcdba69c3295f8991f2).

The architecture object whose sub-objects are renamed is saved to a file as explained in [the previous page](./load_and_save.md#save).

#### `use_latest_message`

Using the following Python API to use `use_latest_message` on an architecture object. The examples below follow the previous section.

- You can add a callback to a publisher with `insert_publisher_callback` function in `Architecture` class.

  - As arguments, the target node name, publishing topic name and publisher callback name must be specified.

- You can update `context_types` to `use_latest_message` between targeted subscription and publisher with `update_message_context` function in `Architecture` class.
  - As arguments, the target node name, subscription topic name and publisher topic name must be specified.

```python
# arch is caret_analyze.architecture.architecture.Architecture-based object

arch.insert_publisher_callback('/pong_node', '/pong', 'timer_callback_1')
arch.update_message_context('/pong_node', '/ping', '/pong', 'use_latest_message')
```

As a result of these processes, the data path is defined as `use_latest_message`.
The edited architecture file is output by `arch.export()` function.

> You can also remove intra-node data path of `use_latest_message` in the Architecture object using the following Python API.

Reviewer 2: This is a sentence I proposed myself, though,,,
It may be "remove" in effect, but the API name is "update", which may confuse first-time users.

How about this? 

```
If you want to disable the intra-node data path, specify "UNDEFINED" in the message context.
```

 I am not confident that this is the correct representation of the actual behavior. So, please check if it is corret.

Author: @nabetetsu 
Fixed it in [e8fcbe0](https://github.com/tier4/CARET_doc/pull/192/commits/e8fcbe0d948128f00440640178e0219bce332df0).

All of the following code snippets can be executed after the `Architecture('type', 'file_path')` method loads an architecture object.
The architecture object whose sub-objects are renamed is saved to a file as explained in [the previous page](./load_and_save.md#save).

#### `use_latest_message`

Using the following Python API to use `use_latest_message` on an architecture object. The examples below follow the previous section.

- You can add callback to publisher with `insert_publisher_callback` function in `Architecture` class.

  - As arguments, the target node name, publishing topic name and publisher callback name must be specified.

- You can update `context_types` to `use_latest_message` between targeted subscription and publisher with `update_message_context` function in `Architecture` class.
  - As arguments, the target node name, subscription topic name and publisher topic name must be specified.

```python
# arch is caret_analyze.architecture.architecture.Architecture-based object

arch.insert_publisher_callback('/pong_node', '/pong', 'timer_callback_1')
arch.update_message_context('/pong_node', '/ping', '/pong', 'use_latest_message')
```

> 

Reviewer 2: You can also remove intra-node data path of `use_latest_message` in Architecture object.

```python
# arch is caret_analyze.architecture.architecture.Architecture-based object
arch.remove_publisher_callback('/pong_node', '/pong', 'timer_callback_1')
arch.update_message_context('/pong_node', '/ping', '/pong', 'UNDEFINED')
```

Author: Fixed it in [ec38045](https://github.com/tier4/CARET_doc/pull/192/commits/ec38045c6dc7894a7a618e76683ed0b16673f01f).
